pkg/icingadb/history/sorter.go

```diff
@@ -0,0 +1,320 @@
+		// struct{}{} would be sent, effectively deadlocking.
+		for len(queueEventCh) > 0 {
+			<-queueEventCh
+		}
+		queueEventCh <- struct{}{}
+	}
+
+	// The actual callback function is executed concurrently as it might block longer than expected. A blocking select
+	// would result in the queue not being populated, effectively blocking the submissionWorker. Thus, the callbackFn is
+	// started in a goroutine, signaling back its success status via callbackCh. If no callback is active, the channel
+	// is nil. Furthermore, an exponential backoff for sequentially failing callbacks is in place.
+	const callbackMaxDelay = 10 * time.Second
+	var callbackDelay time.Duration
+	var callbackCh chan bool
+	callbackFn := func(submission streamSorterSubmission) {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(callbackDelay):
+		}
+
+		start := time.Now()
+		success := sorter.callbackFn(submission.msg, submission.args)
+		if success {
+			submission.out <- submission.msg
+			callbackDelay = 0
+		} else {
+			callbackDelay = min(2*max(time.Millisecond, callbackDelay), callbackMaxDelay)
+		}
+
+		if sorter.verbose {
+			sorter.logger.Debugw("Callback finished",
+				zap.String("id", submission.msg.ID),
+				zap.Bool("success", success),
+				zap.Duration("duration", time.Since(start)),
+				zap.Duration("next-delay", callbackDelay))
+		}
+
+		callbackCh <- success
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case submissions, ok := <-in:
+			if !ok {
+				return
+			}
+
+			queue = append(queue, submissions...)
+			queueEvent()
+
+			if sorter.verbose {
+				sorter.logger.Debugw("Queue worker received new submissions",
+					zap.Int("queue-size", len(queue)))
+			}
+
+		case <-queueEventCh:
+			if len(queue) == 0 {
+				continue
+			}
+
+			if callbackCh != nil {
+				continue
+			}
+			callbackCh = make(chan bool)
+			go callbackFn(queue[0])
+
+		case success := <-callbackCh:
+			if success {
+				queue = queue[1:]
+			}
+
+			close(callbackCh)
+			callbackCh = nil
+
+			if len(queue) > 0 {
+				queueEvent()
+			} else {
+				if sorter.verbose {
+					sorter.logger.Debug("Queue worker finished processing queue")
+				}
+			}
+		}
+	}
+}
+
+// Submit a [redis.XMessage] to the StreamSorter.
+//
+// After the message was sorted and successfully passed to the callback including the optional args, it will be
+// forwarded to the out channel.
+//
+// This method returns an error for malformed Redis Stream IDs or if the internal submission channel blocks for over a
+// second. Usually, this both should not happen.
+func (sorter *StreamSorter) Submit(msg redis.XMessage, args any, out chan<- redis.XMessage) error {
+	ms, seq, err := parseRedisStreamId(msg.ID)
+	if err != nil {
+		return errors.Wrap(err, "cannot parse Redis Stream ID")
+	}
+
+	submission := streamSorterSubmission{
+		msg:          msg,
+		args:         args,
+		out:          out,
+		streamIdMs:   ms,
+		streamIdSeq:  seq,
+		submitTimeNs: time.Now().UnixNano(),
```

Comment:
Why `UnixNano()`? Go's `time.Time` automatically contains a [monotonic value](https://pkg.go.dev/time#hdr-Monotonic_Clocks) that you lose with this call. `submitTimeNs` seems to be only used in a comparison, so just storing the `time.Time` should be the better option for this.